### DISCUSSION TITLE
Error in Heterogeneous graph by applying the PYG model

### DISCUSSION THREAD

**Discussion by amalislam675** [pyg-team/pytorch_geometric#8150] [Q&A]
I try to apply the GCN2Conv model within HeteroGNN, I found this below mentioned error, can you tell how to fix? With the same code, If I apply GCNConv from PYG, it works fine. Here, need to add initial features, that's why, I am giving the x_dict two times as residual feature.
import numpy as np
import torch
from torch_geometric.data import HeteroData

# Random data just to show how store values of nodes work
authors = torch.rand((10,8))
papers = torch.rand((10,8))
authors_y = torch.rand(10)

# Random data just to show how store values of edges work
write_from = torch.tensor(np.random.choice(10, 15, replace = True))
write_to = torch.tensor(np.random.choice(10, 15, replace=True))
write = torch.concat((write_from, write_to)).reshape(-1,15).long()

# Random dat justo to show how store values of edges work
cite_from = torch.tensor(np.random.choice(10, 15, replace=True))
cite_to = torch.tensor(np.random.choice(10, 15, replace=True))
cite = torch.concat((cite_from, cite_to)).reshape(-1,15).long()

edge_weight1 = torch.rand(15)
edge_weight1 = edge_weight1.unsqueeze(-1)
edge_weight2 = torch.rand(15)
edge_weight2 = edge_weight2.unsqueeze(-1)

from torch_geometric.data import HeteroData

data = HeteroData()
data['author'].x = authors
data['paper'].x = papers

data['author', 'writes', 'author'].edge_index=write
data['paper', 'cited_by', 'paper'].edge_index=cite

data['author', 'writes', 'author'].edge_weight= edge_weight1
data['paper', 'cited_by', 'paper'].edge_weight= edge_weight1

print(data)

import torch
import torch_geometric.transforms as T
from torch_geometric.nn import HeteroConv, GCN2Conv, Linear
class HeteroGNN(torch.nn.Module):
    def __init__(self, hidden_channels, out_channels):
        super().__init__()

        self.lin1=torch.nn.Linear(8, hidden_channels)
        self.conv1 = HeteroConv({
                ('author', 'writes', 'author'): GCN2Conv(hidden_channels, alpha=0.5, theta=0.5, layer=1, cached=True),
                ('paper', 'cited_by', 'paper'): GCN2Conv(hidden_channels, alpha=0.5, theta=0.5, layer=1, cached=True)
            }, aggr='sum')
        print(self.conv1)
        self.conv2 = HeteroConv({
                ('author', 'writes', 'author'): GCN2Conv(hidden_channels, alpha=0.5, theta=0.5, layer=2, cached=True),
                ('paper', 'cited_by', 'paper'): GCN2Conv(hidden_channels, alpha=0.5, theta=0.5, layer=2, cached=True)
            }, aggr='sum')

        self.lin = Linear(hidden_channels*2, out_channels)

    def forward(self, x_dict, edge_index_dict, edge_weight_dict):
        x_dict = {key: self.lin1(x) for key, x in x_dict.items()}
        h = self.conv1(x_dict, x_dict, edge_index_dict, edge_weight_dict)
        h = {key: x.relu() for key, x in h.items()}
        
        
        h = self.conv2(h, x_dict, edge_index_dict, edge_weight_dict)
        h = {key: x.relu() for key, x in h.items()}
        
        out = h["author"]
        return out

Error
Traceback (most recent call last):
  File "/123/PYG/example1_1.py", line 87, in <module>
    out = model(data.x_dict, data.edge_index_dict, data.edge_weight_dict)
  File "/123/PYG/.conda/envs/foo/lib/python3.10/site-packages/torch/nn/modules/module.py", line 1130, in _call_impl
    return forward_call(*input, **kwargs)
  File "/123/PYG/example1_1.py", line 67, in forward
    h = self.conv1(x_dict, x_dict, edge_index_dict, edge_weight_dict)
  File "/123/pyg/lib/python3.10/site-packages/torch/nn/modules/module.py", line 1130, in _call_impl
    return forward_call(*input, **kwargs)
  File "/123/pyg/lib/python3.10/site-packages/torch_geometric/nn/conv/hetero_conv.py", line 106, in forward
    src, rel, dst = edge_type
ValueError: too many values to unpack (expected 3)

**Comment by rusty1s** [comment]
You need to call via
```python
h = self.conv1(x_dict, edge_index_dict, edge_weight_dict)
```

**Reply by amalislam675** [reply]
@rusty1s , thankyou for your reply. I am applying GCN2Conv layers which are present here https://pytorch-geometric.readthedocs.io/en/latest/generated/torch_geometric.nn.conv.GCN2Conv.html#torch_geometric.nn.conv.GCN2Conv. Do you think if we pass x_dict, edge_index_dict, edge_weight_dict without passing the residual input tensor, will it work.  Within the forward function of PYG, it is mentioned that x and x0 two tensors should pass.

**Reply by rusty1s** [reply]
Oh, I see. I missed that `GCN2Conv` expects two feature tensors. Let me confirm in a test :)

**Reply by rusty1s** [reply]
Will be fixed in https://github.com/pyg-team/pytorch_geometric/pull/8166.

**Reply by amalislam675** [reply]
@rusty1s , than you for your consideration. As you mentioned, it will fixed in #8166. Through which way, I will use this updated hetroconv module.  I have installed PYG 2.3.1 in my system. Do I need to use this code from source by separately running HetroConv module?

**Reply by rusty1s** [reply]
At best, install PyG from GitHub/nightly (see README for more details).

**Reply by amalislam675** [reply]
@rusty1s , thanks, I try to intall.
